### ISSUE TITLE
Device list tracking does not handle users who leave rooms

### ISSUE THREAD

**Issue by richvdh** [matrix-org/matrix-rust-sdk#1640]
Suppose:
 * We share an encrypted room with Bob
 * Bob leaves that room
 * Bob signs in on a new device
 * Bob rejoins the room
 * We send an encrypted message

In this case, Bob's new device will see a UISI, because we will not encrypt for it.

The problem is that there is no way to mark Bob as 'untracked', so we think we have an up-to-date list of Bob's devices, where we do not. 

I think this is related to https://github.com/matrix-org/matrix-rust-sdk/issues/1191.

**Comment by poljar** [comment]
Interesting, I would have assumed that the server will tell us that Bob has changed at the same time we receive the `m.room.member` event that inserts Bob back into the room.

Why can't it?

**Comment by cvwright** [comment]
Isn't it also a problem that the crypto crate never learned about Bob leaving in the first place?

When somebody leaves (or is kicked/banned), don't you typically want to start a new chain of keys that the left user doesn't know?

**Comment by richvdh** [comment]
I need to investigate this a bit more to figure out exactly what happens, but to do that first I need to fix https://github.com/vector-im/element-web/issues/24795.

**Comment by richvdh** [comment]
> When somebody leaves (or is kicked/banned), don't you typically want to start a new chain of keys that the left user doesn't know?

Yes, but the decision of which users to encrypt for is left to the application ([`share_room_key`](https://docs.rs/matrix-sdk-crypto/0.6.0/matrix_sdk_crypto/struct.OlmMachine.html#method.share_room_key) takes a list of users), so that bit works ok.

**Comment by richvdh** [comment]
> Interesting, I would have assumed that the server will tell us that Bob has changed at the same time we receive the `m.room.member` event that inserts Bob back into the room.

Indeed, it turns out this is exactly what happens. As the [spec](https://spec.matrix.org/v1.6/client-server-api/#a-namee2e-extensions-to-sync-extensions-to-sync) says (my emphasis).

`changed`: List of users who have updated their device identity or cross-signing keys, *or who now share an encrypted room with the client* since the previous sync response.

So, this is a non-issue. Sorry for the noise!

**Comment by richvdh** [comment]
Related: here's where we originally fixed this for the js-sdk back in the day - by adding the user to `changed`: https://github.com/vector-im/element-web/issues/4983
